### ISSUE TITLE
sudden install error with pip (pyparsing 2.1.2 related)

### ISSUE THREAD

**Issue by mdespriee** [matplotlib/matplotlib#6398]
Every day we have a bunch of EC2 machines that start and install themselves.

This morning, the `pip install matplotlib ` returned this :

```
    Complete output from command /usr/bin/python3.4 -c "import setuptools, tokenize;__file__='/tmp/pip-build-y7e201wb/matplotlib/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-1h1al542-record/install-record.txt --single-version-externally-managed --compile:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-y7e201wb/matplotlib/setup.py", line 167, in <module>
        result = package.check()
      File "/tmp/pip-build-y7e201wb/matplotlib/setupext.py", line 1246, in check
        import pyparsing
      File "/usr/local/lib/python3.4/site-packages/pyparsing.py", line 3478, in <module>
        _escapedPunc = Word( _bslash, r"\[]-*.$+^?()~ ", exact=2 ).setParseAction(lambda s,l,t:t[0][1])
      File "/usr/local/lib/python3.4/site-packages/pyparsing.py", line 948, in setParseAction
        self.parseAction = list(map(_trim_arity, list(fns)))
      File "/usr/local/lib/python3.4/site-packages/pyparsing.py", line 808, in _trim_arity
        this_line = extract_stack()[-1]
      File "/usr/local/lib/python3.4/site-packages/pyparsing.py", line 793, in extract_stack
        return [(frame_summary.filename, frame_summary.lineno)]
    AttributeError: 'tuple' object has no attribute 'filename'
    ============================================================================
    Edit setup.cfg to change the build options
    
    BUILDING MATPLOTLIB
                matplotlib: yes [1.5.1]
                    python: yes [3.4.3 (default, Apr  1 2015, 18:10:40)  [GCC
                            4.8.2 20140120 (Red Hat 4.8.2-16)]]
                  platform: yes [linux]
    
    REQUIRED DEPENDENCIES AND EXTENSIONS
                     numpy: yes [version 1.11.0]
                  dateutil: yes [using dateutil version 2.5.3]
                      pytz: yes [using pytz version 2016.4]
                    cycler: yes [using cycler version 0.10.0]
                   tornado: yes [using tornado version 4.3]

```

What happens ? 

Our stack
- python 3.4, amazon linux

**Comment by jenshnielsen** [comment]
I can't reproduce your issue locally but judging from the traceback it looks like this is an issue with pyparsing. I know a new version of pyparsing was just released which pip probably picks up. Could you try installing the second to last version before installing matplotlib?

```
pip install pyparsing==2.1.1
pip install matplotlib
```

**Comment by jenshnielsen** [comment]
Update. Can reproduce it on travis https://travis-ci.org/matplotlib/matplotlib/jobs/129092970

**Comment by jenshnielsen** [comment]
This is the relevant pyparsing issue  https://sourceforge.net/p/pyparsing/bugs/95/ 
Looks like it's a python3.4/3.3 only issue

**Comment by prinsherbert** [comment]
I also stumbled upon this issue, and think this is also a problem for `matplotlib`.

I noticed there are no pinned versions for dependency management. If versions are pinned upon releases or even commits, the committer / releaser has the responsibility to check whether everything works for the specified versions, but also has the voice to disallow certain versions.

PS, I checked for pinning mechanisms by grepping on `pyparsing` in the repository, and couldn't find anything that looked like pinning. So correct me if I missed something.

**Comment by jenshnielsen** [comment]
I have blocked this particular version in #6401 but in general it would be extremely disruptive if matplotlib pinned specific versions of dependencies. Pip/Python in general does not support having more than one version of a dependency installed so consider what happens if you install 2 packages that depend on different versions of a package i.e. pyparsing. If matplotlib depends on version 2.1.1 and package x depends on version 2.1.2. Then running
```
pip install matplotlib 
pip install x
```  
Would result in pyparsing 2.1.1 first being installed and then replaced by 2.1.2 when x is installed. 

Doing it the other way around would result in pyparsing 2.1.1 being installed. 

This means that if you are developing code against pyparsing 2.1.2 and then install matplotlib for something your code might suddenly break since the matplotlib installation has downgraded pyparsing behind your back.

**Comment by jenshnielsen** [comment]
Pyparsing 2.1.3 is out with a fix for this issue

**Comment by skirpichev** [comment]
Are you planing to do a minor matplotlib release with corrected dependencies?  btw, last bugfix release dated Jan 16.

**Comment by tacaswell** [comment]
Closing as a) pyparsing fixed upstream and b) #6401 prevents installation
